pkg/netlink/netlink_watcher.go

```diff
@@ -0,0 +1,1998 @@
+		latestgenmap[k] = v
+	}
+	for k, v := range olddb.FBDB {
+		oldgenmap[k] = v
+	}
+	notify_changes(latestgenmap, oldgenmap, eventStr[2].([]string))
+	// L2Nexthop
+	oldgenmap = make(map[interface{}]interface{})
+	latestgenmap = make(map[interface{}]interface{})
+	for k, v := range latestdb.L2NDB {
+		latestgenmap[k] = v
+	}
+	for k, v := range olddb.L2NDB {
+		oldgenmap[k] = v
+	}
+	notify_changes(latestgenmap, oldgenmap, eventStr[3].([]string))
+}
+
+// resyncWithKernel fun resyncs with kernal db
+func resyncWithKernel() {
+	// Build a new DB snapshot from netlink and other sources
+	readLatestNetlinkState()
+	// Annotate the latest DB entries
+	annotateDBEntries()
+	// Filter the latest DB to retain only entries to be installed
+	applyInstallFilters()
+	// Compute changes between current and latest DB versions and inform subscribers about the changes
+	notifyDBChanges()
+	routes = latestRoutes
+	Nexthops = latestNexthop
+	Neighbors = latestNeighbors
+	fDB = latestFDB
+	l2Nexthops = latestL2Nexthop
+	DeleteLatestDB()
+}
+
+// DeleteLatestDB deletes the latest db snap
+func DeleteLatestDB() {
+	latestRoutes = make(map[routeKey]RouteStruct)
+	latestNeighbors = make(map[neighKey]neighStruct)
+	latestNexthop = make(map[nexthopKey]NexthopStruct)
+	latestFDB = make(map[fdbKey]FdbEntryStruct)
+	latestL2Nexthop = make(map[l2NexthopKey]L2NexthopStruct)
+}
+
+// monitorNetlink moniters the netlink
+func monitorNetlink() {
+	for !stopMonitoring {
+		log.Printf("netlink: Polling netlink databases.")
+		resyncWithKernel()
+		log.Printf("netlink: Polling netlink databases completed.")
+		time.Sleep(time.Duration(pollInterval) * time.Second)
+	}
+	log.Printf("netlink: Stopped periodic polling. Waiting for Infra DB cleanup to finish")
+	time.Sleep(2 * time.Second)
+	log.Printf("netlink: One final netlink poll to identify what's still left.")
+	resyncWithKernel()
+	// Inform subscribers to delete configuration for any still remaining Netlink DB objects.
+	log.Printf("netlink: Delete any residual objects in DB")
+	for _, r := range routes {
+		notifyAddDel(r, "RouteDeleted")
+	}
+	for _, nexthop := range Nexthops {
+		notifyAddDel(nexthop, "NexthopDeleted")
+	}
+	for _, m := range fDB {
+		notifyAddDel(m, "fdb_entry_deleted")
+	}
+	log.Printf("netlink: DB cleanup completed.")
+}
+
+// Initialize function intializes config
+func Initialize() {
+	pollInterval = config.GlobalConfig.Netlink.PollInterval
+	log.Printf("netlink: poll interval: %v", pollInterval)
+	nlEnabled := config.GlobalConfig.Netlink.Enabled
+	if !nlEnabled {
+		log.Printf("netlink: netlink_monitor disabled")
+		return
+	}
+	for i := 0; i < len(config.GlobalConfig.Netlink.PhyPorts); i++ {
+		phyPorts[config.GlobalConfig.Netlink.PhyPorts[i].Name] = config.GlobalConfig.Netlink.PhyPorts[i].Vsi
+	}
+	getlink()
+	ctx = context.Background()
+	nlink = utils.NewNetlinkWrapper()
+	go monitorNetlink() // monitor Thread started
+}
+
+// DeInitialize function handles stops functionality
+func DeInitialize() {
```

Comment:
How is it planned to be used?

Reply:
it is being used in main.go

Reply:
from another thread, right?

If we use a regular bool here, we can have a potential issue with visibility of the change.
So at least we need to use atomic bool
or
More preferable in Go is to pass a context into init which can be canceled in main
```Go
// main
ctx, cancel := context.WithCancel(context.Background())
watcher.Init(ctx)
// some time later, when we need to stop
cancel()

// in watcher, likely in monitor netlink
ticker := time.Tick(1 * time.Second)
for {
    select {
    case <-cancelCtx.Done():
        // do cleanup, we are exiting
        return

    case <-ticker:
       // do your regular monitor stuff
    }
}
```

Reply:
changes bool to atomic bool

Reply:
Good. But let's create a generic issue for graceful shutdown, and we could transition to context usage there